cli/src/merge_tools/builtin.rs

```diff
@@ -237,6 +258,29 @@ fn make_diff_sections(
     Ok(sections)
 }
 
+// We should render the mode section if:
+// 1. The file exists on only one side and is empty.
+// 2. The file existed on both the left and right and the mode actually changed.
+fn should_render_mode_section(left: &FileInfo, right: &FileInfo) -> bool {
+    if left.file_mode == scm_record::FileMode::absent()
+        || right.file_mode == scm_record::FileMode::absent()
+    {
+        // The file should exist on one side.
+        debug_assert!(left.file_mode != right.file_mode);
+        // The file is absent on one side. If it's empty on the other side, then
+        // a content diff section won't be added later. If we don't add the mode
+        // section now, then the file won't be selectable at all. If the file
+        // isn't empty, then we don't need to add a mode section since the user
+        // can select the contents instead and the mode change is implied.
+        // See https://github.com/martinvonz/jj/issues/3016.
+        return matches!(
+            (left.is_empty(), right.is_empty()),
+            (Some(true), None) | (None, Some(true))
+        );
```

Comment:
nit: I would combine this and `.file_mode` comparison as follows, but I'm not sure which one is more readable. To me, the special case is one side is absent and the other side is not empty.
```rust
match (left.is_empty(), right.is_empty()) {
    (Some(false), None) | (None, Some(false)) => false,
    _ => left.file_mode != right.file_mode,
}
```

Reply:
Done. I like that this is much more concise, and I think it's just as understandable alongside a comment.